### ISSUE TITLE
Random failures when executing atom commands on slave

### ISSUE THREAD

**Issue by josephharrington** [box/ClusterRunner#57]
Stack trace from slave log is below. This exception causes the slave to shutdown.

This is worse because the thread that the exception is thrown on (in this case, Build 92 Subjob 11) never reports back to the master that the subjob failed. So the build will be stuck in BUILDING state on the master.

```
[2014-12-19 17:48:02.153] ERROR   Bld92-Sub11     unhandled_excep Unhandled exception handler caught exception.
Traceback (most recent call last):
  File "/home/jenkins/ClusterRunnerBuild/app/util/safe_thread.py", line 18, in run
  File "/usr/local/lib/python3.4/threading.py", line 868, in run
  File "/home/jenkins/ClusterRunnerBuild/app/slave/cluster_slave.py", line 289, in _execute_subjob
  File "/home/jenkins/ClusterRunnerBuild/app/slave/subjob_executor.py", line 89, in execute_subjob
  File "/home/jenkins/ClusterRunnerBuild/app/slave/subjob_executor.py", line 124, in _execute_atom_command
  File "/home/jenkins/ClusterRunnerBuild/app/project_type/git.py", line 233, in execute_command_in_project
  File "/home/jenkins/ClusterRunnerBuild/app/project_type/project_type.py", line 224, in execute_command_in_project
  File "/usr/local/lib/python3.4/subprocess.py", line 959, in communicate
  File "/usr/local/lib/python3.4/subprocess.py", line 1607, in _communicate
  File "/usr/local/lib/python3.4/selectors.py", line 339, in register
  File "/usr/local/lib/python3.4/selectors.py", line 226, in register
  File "/usr/local/lib/python3.4/selectors.py", line 213, in _fileobj_lookup
  File "/usr/local/lib/python3.4/selectors.py", line 39, in _fileobj_to_fd
ValueError: Invalid file object: <_io.BufferedReader name=27>
```

**Comment by michaelb990** [comment]
I'm seeing the same error... Any known workarounds that you have used to recover the ClusterRunner system?

**Comment by josephharrington** [comment]
Hey Michael, we don't have any good workarounds for this right now (beyond redeploying the entire cluster). Thanks for letting us know you're hitting this too -- now we know it's not just us, and we'll double down on fixing this. 

This is still a relatively infrequent occurrence for us (maybe 1 in 100 builds). Does this happen frequently for you? And just to confirm -- are you seeing the exact same stack trace on the slave? Any other details you could provide (e.g., the size of your deployment, if you're seeing any other weirdness, etc.) would also be helpful. Thanks!

**Comment by michaelb990** [comment]
Basically the same trace, I've copied it below. This has happened a few times, and while I don't have those stack traces, I believe it's a similar root cause. Usually I've just tore it down and redeployed. I've been trying this out as a way to run both UI & command line integration tests in parallel to speed up our test execution. From what I've seen, this error seems to have only happened for me when running UI tests in a browser with Xvfb though it may just be a small sample size.
```
[2015-01-03 02:48:36.609] ERROR   Build52-Sub39   unhandled_excep Unhandled exception handler caught exception.
Traceback (most recent call last):
  File "/home/jenkins/ClusterRunnerBuild/app/util/safe_thread.py", line 18, in run
  File "/usr/local/lib/python3.4/threading.py", line 868, in run
  File "/home/jenkins/ClusterRunnerBuild/app/slave/cluster_slave.py", line 249, in _execute_subjob
  File "/home/jenkins/ClusterRunnerBuild/app/slave/subjob_executor.py", line 89, in execute_subjob
  File "/home/jenkins/ClusterRunnerBuild/app/slave/subjob_executor.py", line 124, in _execute_atom_command
  File "/home/jenkins/ClusterRunnerBuild/app/project_type/git.py", line 173, in execute_command_in_project
  File "/home/jenkins/ClusterRunnerBuild/app/project_type/project_type.py", line 199, in execute_command_in_project
  File "/usr/local/lib/python3.4/subprocess.py", line 959, in communicate
  File "/usr/local/lib/python3.4/subprocess.py", line 1607, in _communicate
  File "/usr/local/lib/python3.4/selectors.py", line 339, in register
  File "/usr/local/lib/python3.4/selectors.py", line 226, in register
  File "/usr/local/lib/python3.4/selectors.py", line 213, in _fileobj_lookup
  File "/usr/local/lib/python3.4/selectors.py", line 39, in _fileobj_to_fd
ValueError: Invalid file object: <_io.BufferedReader name=12>
[2015-01-03 02:48:36.610] DEBUG   Build52-Sub39   unhandled_excep Executing teardown callback: <function ServiceSubcommand._write_pid_file.<locals>.remove_pid_file at 0x7f7ffdfb4b70>
[2015-01-03 02:48:36.611] DEBUG   Build52-Sub39   unhandled_excep Executing teardown callback: functools.partial(<bound method EPollIOLoop.add_callback of <tornado.platform.epoll.EPollIOLoop object at 0x7f7ffe7fd9b0>>, callback=<bound method EPollIOLoop.stop of <tornado.platform.epoll.EPollIOLoop object at 0x7f7ffe7fd9b0>>)
[2015-01-03 02:48:36.611] DEBUG   Build52-Sub39   unhandled_excep Executing teardown callback: <bound method ClusterSlave._async_teardown_build of <app.slave.cluster_slave.ClusterSlave object at 0x7f7ffe7e4e48>>
[2015-01-03 02:48:36.612] NOTICE  SlaveTornadoThr subcommand      Slave server was stopped.
[2015-01-03 02:48:36.665] INFO    Build52-Sub39   project_type    ProjectType teardown complete.
[2015-01-03 02:48:36.666] INFO    Build52-Sub39   cluster_slave   Build teardown complete for build 52.
[2015-01-03 02:48:36.678] INFO    Build52-Sub39   cluster_slave   Notifying master to disconnect this slave.
```
I also noticed these last 3 lines seem to have no impact, as the master doesn't seem to know this slave was disconnected. 

We're running version **0.5.70**

**Comment by josephharrington** [comment]
Thanks, Michael. We hope to have a fix out this week. The first goal will be to add robustness to the slave code so that this only fails the build and doesn't bring down the slave. The underlying cause of "Invalid file object" might be a bit harder to track down, but I'll update here with any findings.

**Comment by michaelb990** [comment]
Cool. Thanks for the quick response. I wanted to let you know that after hitting it again, I was able to run `clusterrunner slave` on the node that failed, it reconnected to the master and the system recovered. So, no need to redeploy entirely. Looking forward to hearing more.

**Comment by josephharrington** [comment]
Just to summarize some of my findings in investigating this issue: 
- I never figured out the exact cause of the ValueError exception. I was able to reproduce a similar stack trace by doing: 
```python
from subprocess import Popen, PIPE
p = Popen('ls', shell=True, stdout=PIPE, stderr=PIPE)
p.communicate()  # returns stdout and stderr from ls
p.communicate()  # raises ValueError exception
```
- The above code snippet *might* hint that we're getting this exception because we're trying to use a closed pipe to communicate with a subprocess. This is odd though because our code follows the exact pattern in the Python docs for using communicate's `timeout` argument and catching `TimeoutExpired`.
- The fix I've proposed in PR #68 is more of a workaround than a true fix. It changes our code to use temporary files to cache subprocess output instead of pipes since we have more direct control over the lifespan of tempfiles. I'm not super satisfied with that as a fix, but it should at least eliminate the issue.
